Reader.cpp

```diff
@@ -42,6 +43,12 @@ Reader::Reader(const std::string& fname, size_t buffer_size)
         stream.reset(static_cast<std::istream*>(file_stream));
     }
 
+    if(stream->fail()) {
```

Comment:
Awesome. I like this solution. ++++ to constructors that throw when they can't establish a reasonable internal state. This is very good design.

Reply:
Yeah, I figured that it would be good to catch the problem as early as possible, and the constructors looked like the best place, since if the file they get passed can't be read/created, the instances don't make any sense.